internal/util/observability/observability.go

```diff
@@ -19,34 +19,82 @@ package observability
 
 import (
 	"context"
+	"errors"
 	"time"
 
 	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
 	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
 	otelsdkresource "go.opentelemetry.io/otel/sdk/resource"
 	otelsdktrace "go.opentelemetry.io/otel/sdk/trace"
 	otelsemconv "go.opentelemetry.io/otel/semconv/v1.21.0"
-
-	"github.com/FerretDB/FerretDB/internal/util/must"
+	"go.uber.org/zap"
 )
 
-// SetupOtel sets up OpenTelemetry exporter with fixed values.
-func SetupOtel(service string) func(context.Context) error {
-	exporter := must.NotFail(otlptracehttp.New(
+// OtelConfig is the configuration for OpenTelemetry.
+type OtelConfig struct {
+	Service  string
+	Version  string
+	Endpoint string
+}
+
+// Otel represents the OpenTelemetry system.
+type Otel struct {
+	l  *zap.Logger
+	tp *otelsdktrace.TracerProvider
+}
+
+// NewOtel sets up OTLP exporter and tracer provider.
+func NewOtel(config *OtelConfig, l *zap.Logger) (*Otel, error) {
+	if config.Endpoint == "" {
+		return nil, errors.New("endpoint is required")
+	}
+
+	var exporter *otlptrace.Exporter
+	var err error
+
+	// Exporter and tracer are configured with the particular params on purpose.
+	// We don't want to let them being set through OTEL_* environment variables,
+	// but we set them explicitly.
+	exporter, err = otlptracehttp.New(
 		context.TODO(),
-		otlptracehttp.WithEndpoint("127.0.0.1:4318"),
+		otlptracehttp.WithEndpoint(config.Endpoint),
 		otlptracehttp.WithInsecure(),
-	))
+	)
+	if err != nil {
+		return nil, err
+	}
 
 	tp := otelsdktrace.NewTracerProvider(
 		otelsdktrace.WithBatcher(exporter, otelsdktrace.WithBatchTimeout(time.Second)),
 		otelsdktrace.WithSampler(otelsdktrace.AlwaysSample()),
 		otelsdktrace.WithResource(otelsdkresource.NewSchemaless(
-			otelsemconv.ServiceNameKey.String(service),
+			otelsemconv.ServiceName(config.Service),
+			otelsemconv.ServiceVersion(config.Version),
 		)),
 	)
 
 	otel.SetTracerProvider(tp)
 
-	return tp.Shutdown
+	return &Otel{
+		l:  l,
+		tp: tp,
+	}, nil
+}
+
+// Run runs the OpenTelemetry system.
+func (o *Otel) Run(ctx context.Context) {
```

Comment:
It seems `OpenTelemetry` is already started or at least this function doesn't seem to run it. It does has deferred call to shudown `TracerProvider` though. What does this function supposed to do?

Reply:
Run "thing" until ctx is canceled, then shutdown gracefully. It follows the common approach in #4404. The comment could be improved, though.